### DISCUSSION TITLE
Find groups in molecule beyond SMARTS match

### DISCUSSION THREAD

**Discussion by bertiewooster** [rdkit/rdkit#5830] [Q&A] [answered]
Is there a way to find the groups in a molecule that do NOT match SMARTS? The goal is to identify a common core, then the pendant groups off it. For example, as illustrated by the [SmilesMCStoGridImage utility I posted earlier](https://github.com/rdkit/rdkit/discussions/5751), here is the maximum common substructure for two molecules:

`SmilesMCStoGridImage({"C1CCC1": "core only", "C1CC(O)C1CC": "two pendant groups"})`

![SMARTS-core-only-two_groups](https://user-images.githubusercontent.com/1335522/204685334-7060173b-6793-4ddc-928d-cd9169368761.png)

The common core is a cyclobutane ring. The first molecule has no groups off the core, and the second molecule has two pendant groups, *OH and *CH<sub>2</sub>CH<sub>3</sub>. I'd like a method that produces an empty list `[]` for the first molecule, and `["*O", "*CC"]` for the second molecule.

I got the non-matching atoms by following a [method in the tutorial](https://www.rdkit.org/docs/GettingStartedInPython.html#drawing-molecules):

```
>>> plot, SMARTS, match_molecule, molecules = SmilesMCStoGridImage({"C1CCC1": "small", with_groups: "two groups"}, verbose=True)
>>> hit_ats_set = set(mol_with_groups.GetSubstructMatch(match_molecule))
>>> hit_ats_set
{0, 1, 2, 4}
>>> num_atoms = mol_with_groups.GetNumAtoms()
>>> num_atoms
7
>>> all_atoms_set = set(range(num_atoms))
>>> all_atoms_set
{0, 1, 2, 3, 4, 5, 6}
>>> non_matching_set = all_atoms_set - hit_ats_set
>>> non_matching_set
{3, 5, 6}
>>> for non_matching_atom in non_matching_set:
...     atom_non_matching = mol_with_groups.GetAtomWithIdx(non_matching_atom)
...     print(atom_non_matching.GetAtomicNum())
8
6
6
>>> Draw.MolsToGridImage(molecules, highlightAtomLists=[[], non_matching_set])
```
![SMARTS-non-matching-highlighted](https://user-images.githubusercontent.com/1335522/205192345-1b0aa861-3c02-4568-8f35-3bda96518e74.png)

While this gets the non-matching atoms, it doesn't quite answer the question of what the _groups_ off the core are.

**Comment by Xu-Yanbo** [comment]
I think these steps will work:
1. In a molecule, we replace all matched atom with atom(*)
2. remove those bonds whose both ends are atom(*)
3. now we  get some fragments, then remove those fragments whose atoms are all atom(*)
4. the left fragments are we want.

**Reply by bertiewooster** [reply]
Thanks for suggesting those steps! I will think about how to implement them and try it out.

**Comment by greglandrum** [comment, marked as answer]
@bertiewooster I think what you're looking for can be done easily using `Chem.ReplaceCore`:
```
In [9]: ms = [Chem.MolFromSmiles(s) for s in ('C1CCC1','C1C(O)C(CC)C1','CC1CCC1')]

In [10]: core = ms[0]

In [17]: res = Chem.ReplaceCore(ms[1],core)

In [18]: Chem.MolToSmiles(res)
Out[18]: '[1*]O.[2*]CC'

In [19]: frags = Chem.GetMolFrags(res,asMols=True)

In [20]: [Chem.MolToSmiles(x) for x in frags]
Out[20]: ['[1*]O', '[2*]CC']
```
The dummy atoms indicate attachment points and their isotopes indicate which atom in the core they were connected to.


There's a more sophisticated solution via the RGroup decomposition code:
```
In [8]: from rdkit.Chem import rdRGroupDecomposition

In [9]: ms = [Chem.MolFromSmiles(s) for s in ('C1CCC1','C1C(O)C(CC)C1','CC1CCC1')]

In [10]: core = ms[0]

In [12]: rgd,_ = rdRGroupDecomposition.RGroupDecompose([core],ms)

In [13]: rgd
Out[13]:
[{'Core': <rdkit.Chem.rdchem.Mol at 0x215d66703c0>,
  'R1': <rdkit.Chem.rdchem.Mol at 0x215d6670120>,
  'R2': <rdkit.Chem.rdchem.Mol at 0x215d6670660>},
 {'Core': <rdkit.Chem.rdchem.Mol at 0x215d5f090e0>,
  'R1': <rdkit.Chem.rdchem.Mol at 0x215d8bba260>,
  'R2': <rdkit.Chem.rdchem.Mol at 0x215d8bb8740>},
 {'Core': <rdkit.Chem.rdchem.Mol at 0x215d8bb8d60>,
  'R1': <rdkit.Chem.rdchem.Mol at 0x215d8bb9cb0>,
  'R2': <rdkit.Chem.rdchem.Mol at 0x215d8f862d0>}]

In [14]: rgd,_ = rdRGroupDecomposition.RGroupDecompose([core],ms,asSmiles=True)

In [15]: rgd
Out[15]:
[{'Core': 'C1CC([*:2])C1[*:1]', 'R1': '[H][*:1]', 'R2': '[H][*:2]'},
 {'Core': 'C1CC([*:2])C1[*:1]', 'R1': 'O[*:1]', 'R2': 'CC[*:2]'},
 {'Core': 'C1CC([*:2])C1[*:1]', 'R1': '[H][*:1]', 'R2': 'C[*:2]'}]
```

**Reply by bertiewooster** [reply]
Thanks @greglandrum! I'll try those approaches.

**Comment by bertiewooster** [comment]
`Chem.ReplaceCore` identifies the groups off the core well. But the "isotope" numbers aren't lining up with the atom number on the original molecule. It actually appears that the "isotope" numbers are sequential, that is the first fragment is 1*, the second fragment is 2*, the third fragment is 3*, etc.:

```
from rdkit.Chem import AllChem as Chem
from rdkit.Chem import rdRGroupDecomposition
from rdkit.Chem import Draw

ms = [Chem.MolFromSmiles(s) for s in ('c1cnccc1','c(CCC)1c(O)nc(C)c(C)c(OO)1')]

core = ms[0]

res = Chem.ReplaceCore(ms[1],core)

s = Chem.MolToSmiles(res)

print(f"{s=}")

frags = Chem.GetMolFrags(res,asMols=True)

frags_as_mols = [Chem.MolToSmiles(x) for x in frags]
print(f"{frags_as_mols=}")

for atom in ms[1].GetAtoms():
    print(f"  Atom {atom.GetIdx()} is {atom.GetSymbol()}")
    # For each atom, set the property "atomNote" to index of the atom
    atom.SetProp("atomNote", str(atom.GetIdx()))
Draw.MolsToGridImage([ms[1], *frags], useSVG=True)
```

Output:

```
s='[1*]CCC.[2*]O.[3*]C.[4*]C.[5*]OO'
frags_as_mols=['[1*]CCC', '[2*]O', '[3*]C', '[4*]C', '[5*]OO']
  Atom 0 is C
  Atom 1 is C
  Atom 2 is C
  Atom 3 is C
  Atom 4 is C
  Atom 5 is O
  Atom 6 is N
  Atom 7 is C
  Atom 8 is C
  Atom 9 is C
  Atom 10 is C
  Atom 11 is C
  Atom 12 is O
  Atom 13 is O
```

![ReplaceCore_pyridine](https://user-images.githubusercontent.com/1335522/209251029-3e66db13-8286-4f1e-9f57-50befc55ff82.svg)

The fragments also appear to be sequentially numbered using `rdRGroupDecomposition`:

```
rgd,_ = rdRGroupDecomposition.RGroupDecompose([core],ms,asSmiles=True)
rgd
```

Output:

```
[{'Core': 'n1c([*:3])c([*:2])c([*:1])c([*:5])c1[*:4]',
  'R1': '[H][*:1]',
  'R2': '[H][*:2]',
  'R3': '[H][*:3]',
  'R4': '[H][*:4]',
  'R5': '[H][*:5]'},
 {'Core': 'n1c([*:3])c([*:2])c([*:1])c([*:5])c1[*:4]',
  'R1': 'OO[*:1]',
  'R2': 'CCC[*:2]',
  'R3': 'O[*:3]',
  'R4': 'C[*:4]',
  'R5': 'C[*:5]'}]
```

It would be great if the isotopes indicated which atom in the core they were connected to. Am I missing something? Is there another way to do that?

**Reply by bp-kelley** [reply]
These aren't isotopes, they are atom maps.  I.e. R1 is atom map 1 in the fragment and in the core, this is the [*:1] here.  You can use molzip to recreate the original molecule in this case.

```
res = [{'Core': 'n1c([*:3])c([*:2])c([*:1])c([*:5])c1[*:4]',
  'R1': '[H][*:1]',
  'R2': '[H][*:2]',
  'R3': '[H][*:3]',
  'R4': '[H][*:4]',
  'R5': '[H][*:5]'},
 {'Core': 'n1c([*:3])c([*:2])c([*:1])c([*:5])c1[*:4]',
  'R1': 'OO[*:1]',
  'R2': 'CCC[*:2]',
  'R3': 'O[*:3]',
  'R4': 'C[*:4]',
  'R5': 'C[*:5]'}]

smi = ".".join(res[1].values())
Chem.molzip(Chem.MolFromSmiles(smi))
```

![image](https://user-images.githubusercontent.com/9888416/209456893-cf7992d9-b830-497a-9873-532ac864fb3f.png)

**Reply by bp-kelley** [reply]
The only caveat is that the Core needs to come first here.

**Comment by bertiewooster** [comment]
The [tutorial section Substructure-based transformations](https://www.rdkit.org/docs/GettingStartedInPython.html?highlight=replacecore#substructure-based-transformations) notes:

> By default the sidechains are labeled based on the order they are found. They can also be labeled according by the number of that core-atom they’re attached to

by using `labelByIndex=True` with `Chem.ReplaceCore`. So the results I posted above are correctly labeled.

Using `labelByIndex=True` does change the results, but doesn't seem to give the correct results, including for the example in the tutorial:

```
m1 = Chem.MolFromSmiles('c1c(CCO)ncnc1C(=O)O')
core = Chem.MolFromSmiles('c1cncnc1')
tmp=Chem.ReplaceCore(m1,core,labelByIndex=True)

for atom in m1.GetAtoms():
    atom.SetProp("atomNote", str(atom.GetIdx()))
frags_labelByIndex = Chem.GetMolFrags(tmp,asMols=True)

frags_labelByIndex
[Chem.MolToSmiles(frag) for frag in frags_labelByIndex]
```

Output:

```
  Atom 0 is C
  Atom 1 is C
  Atom 2 is C
  Atom 3 is C
  Atom 4 is O
  Atom 5 is N
  Atom 6 is C
  Atom 7 is N
  Atom 8 is C
  Atom 9 is C
  Atom 10 is O
  Atom 11 is O
['[1*]CCO', '[5*]C(=O)O']
```

The part that's incorrect is, the carboxylic acid fragment `C(=O)O` reports that it came off atom index 5, but that's a nitrogen atom in the aromatic ring. In reality, the carboxylic acid fragment came off atom index 8, a carbon in the aromatic ring. (The alcohol fragment correctly reports that it came off atom index 1, a carbon in the aromatic ring.)

For a visual, here's a molgrid with the molecule (with its atoms numbered by their index) and the fragments:

![tutorial-example-frags](https://user-images.githubusercontent.com/1335522/209451296-25222960-6d36-44b1-81ba-5fc2a5974f3f.svg)

**Reply by bp-kelley** [reply]
The atoms are labeled based on the core, not the molecule (m1 here)

![image](https://user-images.githubusercontent.com/9888416/209456775-b5c8a015-a88d-4026-85fe-77e102a0c4a5.png)

![image](https://user-images.githubusercontent.com/9888416/209456781-b916c5f1-cf3d-4c7a-9598-49deb1360c36.png)

The idea is that with the core and the fragments, you should be able to combine them to recreate the original molecule, i.e. the dummy atom with isotope 1 bonds to atom index 1 in the core.

**Reply by bertiewooster** [reply]
Ah, I understand. Thanks @bp-kelley! I may create a pull request for the tutorial to add such clarification and illustrations.

**Comment by bertiewooster** [comment]
One apparent minor inconsistency with the isotope (core atom index) labeling: It appears that if the core atom index that a fragment was attached to is 0 (zero), Chem.ReplaceCore gives no isotope number, just `*frag` (where `frag` is the SMILES for the fragment) rather than what I would expect, namely `[0*frag]`. Here's an example:

```
m1 = Chem.MolFromSmiles('c1c(O)cnc(N(CCC))c1(C)')
core = Chem.MolFromSmiles('Nc1ccc(O)cn1')
tmp=Chem.ReplaceCore(m1,core,labelByIndex=True)

for atom in core.GetAtoms():
    atom.SetProp("atomNote", str(atom.GetIdx()))
frags_labelByIndex = Chem.GetMolFrags(tmp,asMols=True)

for atom in m1.GetAtoms():
    atom.SetProp("atomNote", str(atom.GetIdx()))
frags_labelByIndex = Chem.GetMolFrags(tmp,asMols=True)

frags_labelByIndex
print([Chem.MolToSmiles(frag) for frag in frags_labelByIndex])
Draw.MolsToGridImage([m1, core, *frags_labelByIndex], legends=['m1', 'core', 'frag a', 'frag b'], useSVG=True)
```

Output--note the propyl fragment `'*CCC'` with no isotope number:

```
['*CCC', '[2*]C']
```

![propryl-fragment-without-isotope-number](https://user-images.githubusercontent.com/1335522/209479065-5a6548f4-1d93-4803-9b5e-712f3e341e44.svg)

Whereas if the core's SMILES is reversed (atoms are given in the opposite order)

```
m1 = Chem.MolFromSmiles('c1c(O)cnc(N(CCC))c1(C)')
core = Chem.MolFromSmiles('n1cc(O)ccc1N')
tmp=Chem.ReplaceCore(m1,core,labelByIndex=True)

for atom in core.GetAtoms():
    atom.SetProp("atomNote", str(atom.GetIdx()))
frags_labelByIndex = Chem.GetMolFrags(tmp,asMols=True)

for atom in m1.GetAtoms():
    atom.SetProp("atomNote", str(atom.GetIdx()))
frags_labelByIndex = Chem.GetMolFrags(tmp,asMols=True)

frags_labelByIndex
print([Chem.MolToSmiles(frag) for frag in frags_labelByIndex])
Draw.MolsToGridImage([m1, core, *frags_labelByIndex], legends=['m1', 'core', 'frag a', 'frag b'], useSVG=True)
```

the output includes an isotope number on the propyl fragment:

```
['[7*]CCC', '[5*]C']
```

![propryl-fragment-with-isotope-number](https://user-images.githubusercontent.com/1335522/209479087-17f7a4ce-5bab-49b9-8ada-194556c5006d.svg)

Is there some rationale for excluding the 0 (zero) as an isotope number? I believe this is the [C++ code for returning the dummy atom numbers as isotopes](https://github.com/rdkit/rdkit/blob/master/Code/GraphMol/ChemTransforms/ChemTransforms.cpp#L655).

**Comment by bp-kelley** [comment]
If you call getisotope on the dummy atom you get 0 which is the index in the core.

**Comment by bertiewooster** [comment]
For those interested, I used the `Chem.ReplaceCore` approach @greglandrum suggested. Here's a blog post with the code: [Find the Maximum Common Substructure, and Groups Off It, For a Set of Molecules Using RDKit](https://bertiewooster.github.io/2022/12/25/RDKit-Find-Groups-Off-Common-Core.html). Here's an example output, annotated to explain the grid arrangement:

![Grid of maximum common substructure and core; molecules and groups off maximum common substructure"](https://bertiewooster.github.io/images/pyridinols-MCS-groups-annotated.png)

Thanks for everyone's help!
